src/sql/src/query_model/scalar_expr.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashSet;
+use std::fmt;
+
+use repr::*;
+
+use crate::query_model::{QuantifierId, QuantifierSet};
+
+/// Representation for scalar expressions within a query graph model.
+/// Similar to HirScalarExpr but:
+/// * subqueries are represented as column references to the subquery
+///   quantifiers within the same box the expression belongs to,
+/// * aggregate expressions are considered scalar expressions here
+///   even though they are only valid in the context of a Grouping box,
+/// * column references are represented by a pair (quantifier ID, column
+///   position),
+/// * BaseColumn is used to represent leaf columns, only allowed in
+///   the projection of BaseTables and TableFunctions.
+#[derive(Debug, PartialEq, Clone)]
+pub enum Expr {
+    ColumnReference(ColumnReference),
+    BaseColumn(BaseColumn),
+    Literal(Row, ColumnType),
+}
+
+impl fmt::Display for Expr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self {
+            Expr::ColumnReference(c) => {
+                write!(f, "Q{}.C{}", c.quantifier_id, c.position)
+            }
+            Expr::BaseColumn(c) => {
+                write!(f, "C{}", c.position)
+            }
+            Expr::Literal(row, _) => {
+                write!(f, "{}", row.unpack_first())
+            }
+        }
+    }
+}
+
+impl Expr {
+    // @todo use a generic visit method
```

Comment:
What do you mean by recursive? It can't contain loops, but it will contain child expressions, as in `HirScalarExpr`.

Reply:
The current definition is:

```rust
pub enum Expr {
    ColumnReference(ColumnReference),
    BaseColumn(BaseColumn),
    Literal(Row, ColumnType),
}
```

It is not recursive, because it does not have variants that accept `Expr`. At the moment a `visit` function for that structure is going to be function that dispatches on the enum variants, but for none of the variants there will be a recursive `visit` call.

Reply:
But it will be recursive for sure very shortly.